src/main/java/pl/edu/mimuw/MainExpression.java

```diff
@@ -0,0 +1,134 @@
+package pl.edu.mimuw;
+
+public class MainExpression extends MathExpression {
+  public MathExpression expression;
+  public static boolean corrupted = false;
```

Comment:
tu jest namieszane bardzo - z jednej strony działasz na statycznych polach klasy (czyli jest jedna instancja na cały JVM i zachowuje sie to jak zmienna globalna) a z drugiej strony tworzysz w jednym miejscu obiekt tej klasy przez `new MainExpression()` czyli działasz na konkretnej instancji a nie globalnych zmiennych - co byłoby, jakby gtos utorzył 2 zmienna lokalną przez `new`?